Source/NSOrderedSet.m

```diff
@@ -0,0 +1,1894 @@
+  NSUInteger c = [self count];
+  NSRange range = NSMakeRange(0,c);
+  
+  GS_BEGINIDBUF(objs, count);
+
+  // Get the indexes
+  [indexes getIndexes: indexArray
+	     maxCount: count
+	 inIndexRange: NULL];
+
+  // Get the objects from this set
+  [self getObjects: objs range: range];
+
+  // Iterate over the indexes and replace the objs
+  for (i = 0; i < count; i++)
+    {
+      id obj = [objects objectAtIndex: i];
+      NSUInteger indx = indexArray[i];
+      objs[indx] = obj;
+    }
+
+  // Remove all objects in this set...
+  [self removeAllObjects];
+
+  // Rebuild set...
+  [self addObjects: objs count: c];
+
+  GS_ENDIDBUF();
+}
+
+- (void) replaceObjectsInRange: (NSRange)aRange
+                   withObjects: (const id[])objects
+                         count: (NSUInteger)count
+{
+  id o = nil;
+  NSUInteger i = count;
+  
+  if (count < (aRange.location + aRange.length))
+    [NSException raise: NSRangeException
+		 format: @"Replacing objects beyond end of const[] id."];
+  [self removeObjectsInRange: aRange];
+
+  while (i-- > 0)
+    {
+      o = objects[i];
+      [self insertObject: o atIndex: aRange.location]; 
+    }
+}
+
+- (void)setObject:(id)anObject atIndex:(NSUInteger)anIndex
+{
+  if ([self count] == anIndex)
+    {
+      [self addObject: anObject];
+    }
+  else
+    {
+      [self replaceObjectAtIndex: anIndex withObject: anObject];
+    }
+}
+
+- (void)moveObjectsAtIndexes:(NSIndexSet *)indexes toIndex:(NSUInteger)index
+{
+  NSUInteger count = [indexes count];
+  NSUInteger i = count;
+  NSUInteger indexArray[count];
+  NSMutableArray *tmpArray = [NSMutableArray arrayWithCapacity: count];
+  id o = nil;
+  NSEnumerator *e = nil;
+  
+  [indexes getIndexes: indexArray
+             maxCount: count
+         inIndexRange: NULL];
+
+  // Build the temporary array....
+  while (i-- > 0)
+    {
+      NSUInteger index = indexArray[i];
+      id obj = [self objectAtIndex: index];
+      [tmpArray addObject: obj];
+    }
+
+    // Remove the originals...  
+  for(i = 0; i < count; i++)
+    {
+      NSUInteger index = indexArray[i];
+      [self removeObjectAtIndex: index];
+    }
+  
+  // Move the objects
+  e = [tmpArray objectEnumerator];
+  while((o = [e nextObject]) != nil)
+    {
+      [self insertObject: o atIndex: index];
+    }
+}
+
+- (void) exchangeObjectAtIndex:(NSUInteger)index
+	     withObjectAtIndex:(NSUInteger)otherIndex
+{
+  NSUInteger count = [self count];
+  
+  GS_BEGINIDBUF(objs, count);
+  if (index >= count)
+    {
+      [self _raiseRangeExceptionWithIndex: index from: _cmd];
+    }
+  if (otherIndex >= count)
+    {
+      [self _raiseRangeExceptionWithIndex: otherIndex from: _cmd];
+    }
+  if (index != otherIndex)
+    {
+      id tmp = nil;
+      NSRange range = NSMakeRange(0,[self count]);
+      
+      [self getObjects: objs range: range]; 
+      tmp = objs[index];
+      objs[index] = objs[otherIndex];
+      objs[otherIndex] = tmp;
+      
+      [self removeAllObjects];
```

Comment:
Again, this is very inefficient. We could just remove the object at the higher index first, then the one at the lower and insert them in the opposite order.

Reply:
Implemented as you suggested.